# QuickVariants

## Code for paper "Fast and Accurate Variant Identification Tool for Sequencing-Based Studies"

QuickVariants is a fast and accurate variant identification tool, designed to summarize allele information from read alignments without discarding or filtering the data.

### Usage:

Execute the following command:

```
java -Xms10g -Xmx10g -jar quick-variants.jar [--out-vcf <out.vcf>] [--out-mutations <out.txt>] --reference <ref.fasta> --in-sam <input.sam> --num-threads num_threads [options]
```

This command converts a SAM file to other formats, most notably .vcf.

#### INPUT:
- `--reference <file>`: The reference to use. Should be in .fasta/.fa/.fna or .fastq/.fq format or a .gz of one of those formats.

#### OUTPUT FORMATS:
Summary by reference position, mutation, genome, and raw output are possible.

**Summary by reference position**:
- Options for output files, mutation counts, exclusion of non-mutations, and more.

**Summary by mutation**:
- Includes options for listing mutations, detecting indels, and setting thresholds for various features.

**Summary by genome**:
- `--out-refs-map-count <file>`: Output file to summarize the number of reads mapped to each combination of references.

**Raw output**:
- Includes options for SAM format and unaligned reads, as well as verbosity settings for debugging.

Multiple output formats may be specified during a single run.

#### OTHER:
- `--num-threads <count>`: Number of threads to use for processing.

### Additional Scripts and Models
The `benchmark_scripts` folder contains code used to construct the benchmark dataset, filter SNPs and indels in VCF files, and analyze VCF files.

- `SNP_model.py`: Insert in silico mutations and indels randomly into reference genomes, and generate alignment code.
- `SNPfilter.py`: Filter SNPs in VCF files generated by QuickVariants and bcftools.
- `Indelfiter.py`: Filter indels in VCF files generated by QuickVariants and bcftools, and summarize the number of false positives and false negatives for each dataset.
- `SNP_model_compare.py`: Summarize the number of false positive and false negative SNPs for each dataset.
- `VCFAnalysis.ipynb`: Code to analyze VCF files and visualize results in the samtovcf folder.
